zerver/tests/test_auth_backends.py

```diff
@@ -1891,6 +1891,65 @@ def test_validate_state(self) -> None:
             self.assertIn('login', result.url)
             m.assert_called_with("Sign in with Apple failed: bad state token.")
 
+class AppleAuthBackendNativeFlowTest(AppleIdAuthBackendTest):
+    SIGNUP_URL = '/complete/apple/'
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         desktop_flow_otp: Optional[str]=None,
+                         is_signup: bool=False,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         alternative_start_url: Optional[str]=None,
+                         user_agent: Optional[str]=None,
+                         **extra_data: Any) -> HttpResponse:
+        url, headers = self.prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key, alternative_start_url=self.AUTH_FINISH_URL,
+            user_agent=user_agent,
+        )
+
+        params = {'native_flow': 'true'}
+
+        if not extra_data.get('skip_id_token'):
+            params['id_token'] = self.generate_id_token(account_data_dict)
+
+        if is_signup:
+            params['is_signup'] = '1'
+
+        if subdomain:
+            params['subdomain'] = subdomain
+
+        url += "&%s" % (urllib.parse.urlencode(params),)
+
+        with responses.RequestsMock(assert_all_requests_are_fired=False) as requests_mock:
+            # The below mock is to get public key we generated
+            # instead of getting it from
+            # https://appleid.apple.com/auth/keys.
+            requests_mock.add(
+                requests_mock.GET,
+                self.BACKEND_CLASS.JWK_URL,
+                status=200,
+                json=json.loads(settings.APPLE_JWK),
+            )
+            result = self.client_get(url, **headers)
+        return result
+
+    def test_no_id_token_sent(self) -> None:
+        account_data_dict = self.get_account_data_dict(email=self.email, name=self.name)
+        result = self.social_auth_test(account_data_dict,
+                                       expect_choose_email_screen=False,
+                                       subdomain='zulip', next='/user_uploads/image',
+                                       skip_id_token=True)
+        self.assert_json_error(result, "Missing id_token parameter")
+
+    def test_social_auth_desktop_success(self) -> None:
+        pass
+
+    def test_social_auth_no_key(self) -> None:
+        pass
```

Comment:
Can you add comments on these two disabled tests explaining why we're doing that for the native flow?

Reply:
added.

Reply:
I don't understand why that means we can't / shouldn't run these tests; can you explain?

Reply:
These tests get the `result` from the following `if` in `social_auth_test` and do not reach the end where `AUTH_FINISH_URL` is called
```
if result.status_code != 302 or not result.url.startswith(expected_result_url_prefix):
            return result
```
Since we skip all of that code here in the native flow test, we cannot expect those tests to work.